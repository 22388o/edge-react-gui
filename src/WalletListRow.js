// @flow

import { type EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { Text, View } from 'react-native'
import { TouchableOpacity } from 'react-native-gesture-handler'
import { cacheStyles } from 'react-native-patina'

import { useTheme } from './components/services/ThemeContext'
import { CurrencyIcon } from './CurrencyIcon'
import { useEffect, useState } from './types/reactHooks'
import { type Theme } from './types/Theme'

type Props = {
  wallet: EdgeCurrencyWallet,
  tokenId: string,
  onPress?: () => void
}

export function WalletListRow(props: Props) {
  const { wallet, tokenId, onPress } = props
  const theme = useTheme()
  const styles = getStyles(theme)

  // Look up wallet info:
  const { customTokens = {}, builtinTokens = {} } = wallet.currencyConfig
  const token = customTokens[tokenId] ?? builtinTokens[tokenId]
  const currencyCode = token == null ? wallet.currencyInfo.currencyCode : token.currencyCode
  const [denomination] = token == null ? wallet.currencyInfo.denominations : token.denominations

  // Subscribe to the name:
  const [name, setName] = useState(wallet.name)
  useEffect(() => wallet.watch('name', setName), [wallet])

  // Subscribe to the balance:
  const [balance, setBalance] = useState(wallet.balances[currencyCode] ?? '')
  useEffect(() => wallet.watch('balances', balances => setBalance(balances[currencyCode] ?? '')), [currencyCode, wallet])

  const cryptoBalance = denomination.symbol == null ? balance : `${denomination.symbol} ${balance}`

  return (
    <TouchableOpacity style={styles.container} onPress={onPress}>
      <CurrencyIcon wallet={wallet} tokenId={tokenId} marginRem={0.5} />
      {/* <Text style={styles.iconText}>{`${(syncRatio * 100).toFixed(2)}%`}</Text> */}

      <View style={styles.column}>
        <View style={styles.row}>
          <Text style={styles.currencyText}>{currencyCode}</Text>
          <Text style={styles.tickerText}>Price</Text>
          <Text style={styles.balanceText}>{cryptoBalance}</Text>
        </View>
        <View style={styles.row}>
          <Text style={styles.nameText}>{name ?? `My ${wallet.currencyInfo.displayName}`}</Text>
          <Text style={styles.fiatBalanceText}>Much $$</Text>
        </View>
      </View>
    </TouchableOpacity>
  )
}

const getStyles = cacheStyles((theme: Theme) => {
  const topText = {
    color: theme.primaryText,
    flexShrink: 1,
    fontFamily: theme.fontFaceDefault,
    fontSize: theme.rem(1)
  }
  const bottomText = {
    color: theme.secondaryText,
    flexShrink: 1,
    fontFamily: theme.fontFaceDefault,
    fontSize: theme.rem(0.75)
  }

  return {
    container: {
      alignItems: 'center',
      flexDirection: 'row',
      height: theme.rem(4.25),
      justifyContent: 'flex-start',
      padding: theme.rem(0.5),

      backgroundColor: theme.backgroundGradientRight
    },
    column: {
      alignItems: 'stretch',
      // borderColor: 'red',
      // borderWidth: 1,
      flex: 1,
      flexDirection: 'column',
      justifyContent: 'center',
      padding: theme.rem(0.5)
    },
    row: {
      flexDirection: 'row',
      justifyContent: 'space-between'
    },
    currencyText: { ...topText, fontFamily: theme.fontFaceMedium },
    tickerText: {
      ...topText,
      color: theme.negativeText,
      flexGrow: 1,
      marginLeft: theme.rem(0.75)
    },
    balanceText: { ...topText },
    fiatBalanceText: { ...bottomText },
    nameText: { ...bottomText, flexGrow: 1 },
    iconText: {
      color: theme.secondaryText,
      flexShrink: 1,
      fontFamily: theme.fontFaceDefault,
      fontSize: theme.rem(0.75),
      width: theme.rem(5)
    }
  }
})
