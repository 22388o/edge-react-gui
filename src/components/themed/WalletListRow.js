// @flow

import * as React from 'react'
import { ActivityIndicator, TouchableOpacity, View } from 'react-native'
import FastImage from 'react-native-fast-image'

import { Gradient } from '../../modules/UI/components/Gradient/Gradient.ui.js'
import { memo } from '../../types/reactHooks.js'
import { type Theme, cacheStyles, useTheme } from '../services/ThemeContext.js'
import { EdgeText } from './EdgeText.js'

type Props = {
  currencyCode: string,
  exchangeRateText?: string,
  exchangeRateType?: 'neutral' | 'positive' | 'negative',
  children?: React.Node,
  icon?: React.Node,
  iconUri?: string,
  iconSizeRem?: number,
  gradient?: boolean,
  onPress?: () => void,
  onLongPress?: () => void,
  walletName?: string
}

export const WalletListRowComponent = (props: Props) => {
  const { currencyCode, children, gradient = false, icon, iconUri = '', iconSizeRem, onPress, onLongPress, walletName = '' } = props
  const theme = useTheme()
  const styles = getStyles(theme)

  const iconSizeStyle = iconSizeRem != null ? { width: theme.rem(iconSizeRem), height: theme.rem(iconSizeRem) } : styles.iconSize
  const iconComponent = icon == null ? <FastImage style={iconSizeStyle} source={{ uri: iconUri }} /> : icon

  const WalletRow = ({ children }) =>
    gradient ? <Gradient style={styles.containerGradient}>{children}</Gradient> : <View style={styles.container}>{children}</View>

  return (
    <WalletRow gradient={gradient}>
      <TouchableOpacity onPress={onPress} onLongPress={onLongPress}>
        {currencyCode === '' ? (
          <View style={styles.loaderContainer}>
            <ActivityIndicator color={theme.primaryText} size="large" />
          </View>
        ) : (
          <View style={styles.contentContainer}>
            <View style={styles.iconContainer}>{iconComponent}</View>
            <View style={styles.detailsContainer}>
              <View style={styles.detailsTop}>
                <EdgeText style={styles.detailsCurrency}>{currencyCode}</EdgeText>
                {props.exchangeRateText != null ? (
                  <EdgeText style={[styles.exchangeRateStyle, styles[props.exchangeRateType ?? 'neutral']]}>{props.exchangeRateText}</EdgeText>
                ) : null}
              </View>
              <View style={styles.detailsBottom}>
                <EdgeText style={styles.detailsName}>{walletName}</EdgeText>
              </View>
            </View>
            <View style={styles.childrenContainer}>{children}</View>
          </View>
        )}
      </TouchableOpacity>
    </WalletRow>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  // Background
  containerGradient: {
    flex: 1,
    paddingHorizontal: theme.rem(1)
  },
  container: {
    backgroundColor: theme.modal
  },
  // Top level Containers
  contentContainer: {
    flexDirection: 'row',
    marginVertical: theme.rem(1)
  },
  loaderContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    height: theme.rem(4.25),
    paddingHorizontal: theme.rem(1.75)
  },
  // Data containers //
  // Icon Container
  iconContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    marginRight: theme.rem(1)
  },
  // Details Container
  detailsContainer: {
    flex: 1,
    flexDirection: 'column',
    marginRight: theme.rem(0.5)
  },
  // Children (Right part) Container
  childrenContainer: {
    flex: 0
  },
  // Other styles
  iconSize: {
    width: theme.rem(2),
    height: theme.rem(2)
  },
  detailsTop: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start'
  },
  detailsBottom: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start'
  },
  detailsCurrency: {
    flexBasis: 'auto',
    flexShrink: 1,
    fontFamily: theme.fontFaceMedium
  },
  exchangeRateStyle: {
    textAlign: 'left',
    flexBasis: 'auto',
    flexShrink: 1,
    marginLeft: theme.rem(0.75)
  },
  detailsName: {
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  },
  // Difference Percentage Styles
  neutral: {
    color: theme.secondaryText
  },
  positive: {
    color: theme.positiveText
  },
  negative: {
    color: theme.negativeText
  }
}))

export const WalletListRow = memo(WalletListRowComponent)
