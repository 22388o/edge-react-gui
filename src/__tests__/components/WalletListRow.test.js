/* globals describe it expect */
/* eslint-disable flowtype/require-valid-file-annotation */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { WalletListRow } from '../../components/themed/WalletListRow.js'

describe('WalletListRow', () => {
  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      currencyCode: 'string',
      children: 'Hello',
      icon: 'btc',
      editIcon: 'btc',
      gradient: true,
      onPress: () => undefined,
      onLongPress: () => undefined,
      walletName: 'My bitcoin wallet'
    }
    const actual = renderer.render(<WalletListRow {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
