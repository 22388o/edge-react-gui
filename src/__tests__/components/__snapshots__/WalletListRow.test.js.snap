// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletListRow should render with loading props 1`] = `
<WalletRow
  gradient={true}
>
  <TouchableOpacity
    onLongPress={[Function]}
    onPress={[Function]}
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginVertical": 22,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginRight": 22,
          }
        }
      >
        btc
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "column",
            "marginRight": 11,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "flex-start",
            }
          }
        >
          <WithTheme(EdgeTextComponent)
            style={
              Object {
                "flexBasis": "auto",
                "flexShrink": 1,
                "fontFamily": "Quicksand-Medium",
              }
            }
          >
            string
          </WithTheme(EdgeTextComponent)>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "flex-start",
            }
          }
        >
          <WithTheme(EdgeTextComponent)
            style={
              Object {
                "color": "#A4C7DF",
                "fontSize": 17,
              }
            }
          >
            My bitcoin wallet
          </WithTheme(EdgeTextComponent)>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 0,
          }
        }
      >
        Hello
      </View>
    </View>
  </TouchableOpacity>
</WalletRow>
`;
